File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSSystemMessageCell.m
//
//  MSSystemMessageCell.m
//  BlackFireIM
//
//  Created by benny wang on 2021/3/22.
//

#import "MSSystemMessageCell.h"
#import "UIColor+BFDarkMode.h"


@interface MSSystemMessageCell()

@property(nonatomic,strong) UILabel *messageLabel;

@property(nonatomic,strong) MSSystemMessageCellData *systemData;

@end
@implementation MSSystemMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        _messageLabel = [[UILabel alloc] init];
        _messageLabel.font = [UIFont systemFontOfSize:13];
        _messageLabel.textColor = [UIColor d_systemGrayColor];
        _messageLabel.textAlignment = NSTextAlignmentCenter;
        _messageLabel.numberOfLines = 0;
        _messageLabel.backgroundColor = [UIColor clearColor];
        _messageLabel.layer.cornerRadius = 3;
        [_messageLabel.layer setMasksToBounds:YES];
        [self.container addSubview:_messageLabel];
    }
    return self;
}

- (MSSystemMessageCellData *)systemData
{
    return (MSSystemMessageCellData *)self.messageData;
}

- (void)fillWithData:(MSSystemMessageCellData *)data;
{
    [super fillWithData:data];
    self.messageLabel.text = data.content;
    self.messageLabel.textColor = data.contentColor;
    self.nameLabel.hidden = YES;
    self.avatarView.hidden = YES;
    self.retryView.hidden = YES;
    [self.indicator stopAnimating];
    [self setNeedsLayout];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.container.center = self.container.superview.center;
    self.messageLabel.frame = self.container.bounds;
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSSystemMessageCell.swift
//
//  MSSystemMessageCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/10.
//

import UIKit

open class MSSystemMessageCell: MSMessageCell {

    public private(set) var messageLabel: UILabel!
    
    public private(set) var systemData: MSSystemMessageCellData!
    
    public override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        messageLabel = UILabel()
        messageLabel.font = .systemFont(ofSize: 13)
        messageLabel.textColor = .systemGray
        messageLabel.textAlignment = .center
        messageLabel.numberOfLines = 0
        messageLabel.backgroundColor = .clear
        messageLabel.layer.cornerRadius = 3
        messageLabel.layer.masksToBounds = true
        container.addSubview(messageLabel)
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    public override func fillWithData(data: MSMessageCellData) {
        super.fillWithData(data: data)
        if let systemData = data as? MSSystemMessageCellData {
            self.systemData = systemData
            
            messageLabel.text = systemData.content
            messageLabel.textColor = systemData.contentColor
            nameLabel.isHidden = true
            avatarView.isHidden = true
            retryView.isHidden = true
            indicator.stopAnimating()
            setNeedsLayout()
        }
    }
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        container.center = container.superview!.center
        messageLabel.frame = container.bounds
    }
}
